SAI/sai_api_gen.py

```diff
@@ -0,0 +1,241 @@
+
+        if sai_table_data['name'] in ignore_tables:
+            continue
+
+        for key in table[KEY_TAG]:
+            sai_table_data['keys'].append(get_sai_key_data(program, key))
+
+        for action in table['actions']:
+            if action != 'NoAction':
+                sai_table_data['actions'].append(get_sai_action_data(program, action))
+
+        if len(sai_table_data['keys']) == 1 and sai_table_data['keys'][0]['sai_key_name'] == (table_name.split('.')[-1] + '_id'):
+            sai_table_data['is_object'] = 'true'
+            # Object ID itself is a key
+            sai_table_data['keys'] = []
+        elif len(sai_table_data['keys']) > 5:
+            sai_table_data['is_object'] = 'true'
+        else:
+            sai_table_data['is_object'] = 'false'
+            sai_table_data['name'] = sai_table_data['name'] + '_entry'
+
+        sai_tables.append(sai_table_data)
+
+    sai_api['tables'] = sai_tables
+    return sai_api
+
+
+def write_sai_files(sai_api):
+    # The main file
+    with open('saiapi.h.j2', 'r') as sai_header_tm_file:
+        sai_header_tm_str = sai_header_tm_file.read()
+
+    env = Environment(loader=FileSystemLoader('.'), trim_blocks=True, lstrip_blocks=True)
+    sai_header_tm = env.get_template('saiapi.h.j2')
+    sai_header_str = sai_header_tm.render(sai_api = sai_api)
+
+    with open('./SAI/experimental/saiexperimental' + sai_api['app_name'] + '.h', 'w') as o:
+        o.write(sai_header_str)
+
+    # The SAI Extensions
+    with open('./SAI/experimental/saiextensions.h', 'r') as f:
+        lines = f.readlines()
+
+    new_lines = []
+    for line in lines:
+        if 'Add new experimental APIs above this line' in line:
+            new_lines.append('    SAI_API_' + sai_api['app_name'].upper() + ',\n\n')
+
+        new_lines.append(line)
+
+    with open('./SAI/experimental/saiextensions.h', 'w') as f:
+        f.write(''.join(new_lines))
+
+    # The SAI Type Extensions
+    with open('./SAI/experimental/saitypesextensions.h', 'r') as f:
+        lines = f.readlines()
+
+    new_lines = []
+    for line in lines:
+        if 'Add new experimental object types above this line' in line:
+            for table in sai_api['tables']:
+                new_lines.append('    SAI_OBJECT_TYPE_' + table['name'].upper() + ',\n\n')
+
+        new_lines.append(line)
+
+    with open('./SAI/experimental/saitypesextensions.h', 'w') as f:
+        f.write(''.join(new_lines))
+
+
+# CLI
+parser = argparse.ArgumentParser(description='P4 SAI API generator')
+parser.add_argument('filepath', type=str, help='Path to P4 program BMV2 JSON file')
+parser.add_argument('apiname', type=str, help='Name of the new SAI API')
+parser.add_argument('--print-sai-lib', type=bool)
+parser.add_argument('--sai-git-url', type=str, default='https://github.com/Opencomputeproject/SAI')
+parser.add_argument('--ignore-tables', type=str, default='', help='Coma separated list of tables to ignore')
```

Comment:
Can you provide a default list per above comment?

Reply:
Default is an empty list. I provided tables to ignore in the example for sirius pipeline.